src/rpc/src/chainStorageWatcher.ts

```diff
@@ -0,0 +1,167 @@
+import { AgoricChainStoragePathKind } from './types';
+import { batchVstorageQuery } from './batchQuery';
+import type { UpdateHandler } from './types';
+import type { Unserialize } from '@endo/marshal';
+
+type Subscriber<T> = {
+  onUpdate: UpdateHandler<T>;
+};
+
+const defaultNewPathQueryDelayMs = 20;
+const defaultRefreshLowerBoundMs = 2000;
+const defaultRefreshUpperBoundMs = 4000;
+
+const randomRefreshPeriod = (
+  refreshLowerBoundMs: number,
+  refreshUpperBoundMs: number,
+) =>
+  Math.floor(Math.random() * (refreshUpperBoundMs - refreshLowerBoundMs)) +
+  refreshLowerBoundMs;
+
+const makePathSubscriber = <T>(onUpdate: UpdateHandler<T>) => ({
+  onUpdate,
+});
+
+/**
+ * Periodically queries the most recent data from chain storage, batching RPC
+ * requests for efficiency.
+ *
+ * @param rpcAddr RPC server URL
+ * @param unserialize CapData unserializer to use
+ * @returns
+ */
+export const makeAgoricChainStorageWatcher = (
+  rpcAddr: string,
+  unserialize: Unserialize<string>,
+  newPathQueryDelayMs = defaultNewPathQueryDelayMs,
+  refreshLowerBoundMs = defaultRefreshLowerBoundMs,
+  refreshUpperBoundMs = defaultRefreshUpperBoundMs,
+) => {
+  // Map of paths to [identifier, value] pairs of most recent response values.
+  //
+  // The 'identifier' is used to avoid notifying subscribers for already-seen
+  // values. For 'data' queries, 'identifier' is the blockheight of the
+  // response. For 'children' queries, 'identifier' is the stringified array
+  // of children.
+  const latestValueCache = new Map<string, [string, unknown]>();
+
+  const watchedPathsToSubscribers = new Map<string, Set<Subscriber<unknown>>>();
```

Comment:
nit: each Subscriber is a different type, depending on the string key. so this isn't really a JS Map of the same value with arbitrary keys. It's more of an map-as-object mapping specific keys to specific values.

If you made it an object you could have a type for it that mapped each vstorage key to its subscriber type. E.g.
```
const watchedPathsToSubscribers = {} as {
  'vaultFactory.metrics'?: Set<Subscriber<VaultFactoryMetricsNotification>>,
};
```
Since that would get repetitive, there could be a named type like, 
```
type vstorageSchema = {
 'vaultFactory.metrics': VaultFactoryMetricsNotification;
}
```
And then this mapping could be generated from that type.

Not at all a blocker :)  But consider factoring this now as an object instead of a map, so that adopting this idea would not require runtime changes.